### DISCUSSION TITLE
WebRTC app cannot connect, no ICE candidates created in Bromite

### DISCUSSION THREAD

**Discussion by mehrvarz** [bromite/bromite#1677] [General]
Hi. This is a WebRTC "answering machine": https://timur.mobi/user/answie

When you open the link, you should hear a quality sound sample (spoken words with up to 280kbps via Opus codec). This works well with Chromium on mac/linux/android, kiwi on android, ff and fennec on android, safari on iphone, ff on mac/linux, ungoogle-chromium on mac/linux. 

But it does not work with Bromite 96.0.4664.54 (and other versions) on android. And also not with ungoogle-chromium on android. In these two browsers I never see connectionstatechange "connected". The process gets stuck after connectionstatechange "connecting". At this point "offer", "answer" and lots of candidates have been exchanged. But the final connect state is never reached. I literally run the same URL in kiwi and ff and fennec on the same android device and it always works. Then I run the URL in Bromite and ungoogled-chromium (again on the same Android device) and it does not connect. Do you have an idea what might go wrong?
Thank you very much.

**Comment by mehrvarz** [comment]
My WebRTC caller on Chromium on Android
```
08:18:45.038 new RTCPeerConnection
08:18:45.048 peerCon addTrack local mute audio input
08:18:45.059 peerCon.createOffer 
08:18:45.068 peerCon onnegotiationneeded createOffer 
08:18:45.095 peerCon onsignalingstate have-local-offer 
08:18:45.096 onnegotiationneeded send callerOffer via ws
08:18:45.098 peerCon onicegatheringstatechange gathering
08:18:45.107 onIce callerCandidate via wsSend 192.168.3.166
08:18:45.110 onIce callerCandidate via wsSend 192.168.3.166
08:18:45.218 onIce callerCandidate via wsSend xxx,yyy.195.49
08:18:45.225 onIce callerCandidate via wsSend xxx,yyy.195.49
08:18:45.234 calleeAnswer setLocalDescription
08:18:45.241 calleeAnswer setRemoteDescription
...
```

Same caller on Bromite on Android (first 7 lines are identical)

```
08:54:34.993 new RTCPeerConnection
08:54:35.008 peerCon addTrack local mute audio input
08:54:35.014 peerCon.createOffer
08:54:35.024 peerCon onnegotiationneeded createOffer
08:54:35.049 peerCon onsignalingstate have-local-offer
08:54:35.051 onnegotiationneeded send callerOffer via ws
08:54:35.055 peerCon onicegatheringstatechange gathering
08:54:35.152 peerCon onicegatheringstatechange complete      <--- change
08:54:35.154 onIce end of candidates
08:54:35.201 signaling receive calleeAnswer
08:54:35.201 calleeAnswer setLocalDescription 
08:54:35.209 calleeAnswer setRemoteDescription
...
```
In Bromite no candidates are being created for the caller. What am I doing wrong? What could lead to this?

**Comment by mehrvarz** [comment]
Could this be somehow STUN/TURN related? I am now looking at the turn logs and there does not seem any access at 08:54 (time of the 2nd log), but there was at 08:18. I am running a turn server with a none standard port 3739. Could this port be blocked in Bromite for some reason?

**Comment by csagan5** [comment]
Possibly due to https://github.com/bromite/bromite/blob/96.0.4664.54/build/patches/Change-default-webRTC-policy-to-not-use-any-address.patch

**Reply by mehrvarz** [reply]
Thank you for your quick reply especially at this time of year!! I wasn't able to find this patch from just looking at the symptoms. I read some discussions and start to understand why ungoogled-chr works on desktop, but also fails on Android (in this regard).

I have to say this doesn't look 100% WebRTC complient to me:

```
port_config.enable_multiple_routes = false;
port_config.enable_nonproxied_udp = false;
```
It certainly goes beyond "ungoogled". You said somewhere that if an app does not work by default...

> the user might have to change a flag.

A flag? Or better yet, can I fix this in code, so the code can work everywhere? My desired target platform in this case is Android Webview, but the JS code should ideally be portable.

**Reply by csagan5** [reply]
> I have to say this doesn't look 100% WebRTC compliant to me:

Please elaborate on which WebRTC standard is violated.
You can check the linked issue in the patch for more information on why this patch exists.

> You said somewhere that if an app does not work by default...

What comment are you quoting, specifically? It is otherwise out of context.

> can I fix this in code, so the code can work everywhere?

First you should check why the patch was introduced; if modifications to this patch would weaken privacy then such modifications will not be used in Bromite, but you can of course use it in your own built SystemWebView.

**Reply by mehrvarz** [reply]
> First you should check why the patch was introduced; if modifications to this patch would weaken privacy then such modifications will not be used in Bromite, but you can of course use it in your own built SystemWebView.

If you need to use Tor/VPN to hide your IP, having WebRTC in the browser is a bad thing.

If you want to do p2p realtime media streaming, then not having WebRTC is a bad thing.

Right now, no browser (and no SystemWebView) exists for Android, that supports WebRTC and is also ungoogled.

My Q is: Why not offer two variants? 1 that supports WebRTC 100% and 1 that blocks it 100%. This would be better than claiming support for WebRTC and protecting from it at the same time. Which I don't think is possible.

In case of System WebView: users could even install both variants and switch between them with relative ease. That would be really nice.

**Comment by mehrvarz** [comment]
The WebRTC-policy patch cripples WebRTC in Bromite so that no P2P calls can be made. Users may get the impression that the issue is with the app they are using. WebCall software (link in OP) can now detect the issue and will report that it is caused by the browser.
